File: beeproblem/beeproblem.cpp
#include <bits/stdc++.h>

typedef long long ll;
using namespace std;

int fill(int i, int j, int n, int m, vector<vector<char>>& cells) {
    int count = 1, s = -1, b = 1;

    if (i % 2) {
        s = 0;
        b = 2;
    }
    
    // top bottom
    for (int x = s; x < b; x++) {
        for (int y = -1; y < 2; y += 2) {
            if ( 0 > j + x || 0 > i + y || 
                j + x >= m || i + y >= n || cells[i + y][j + x] == '#') continue;
            
            cells[i + y][j + x] = '#';
            count += fill(i + y, j + x, n, m, cells);
        }
    }
    // left right
    for (int x = -1; x < 2; x++) {
        if (0 > j + x || j + x >= m || cells[i][j + x] == '#') continue;
        
        cells[i][j + x] = '#';
        count += fill(i, j + x, n, m, cells);
    }

    return count;
}

int main() {
    int h, n, m;
    cin >> h >> n >> m;

    vector<vector<char>> cells(n, vector<char>(m));

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            cin >> cells[i][j];
        }
    }

    vector<int> counts;

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (cells[i][j] == '.') {
                cells[i][j] = '#';
                counts.push_back(fill(i, j, n, m, cells));
            }
        }
    }

    sort(counts.begin(), counts.end());

    int fills = 0;
    while (h > 0) {
        h -= counts.back();
        counts.pop_back();
        fills++;
    }

    cout << fills << endl;

    return 0;
}

File: beeproblem/old/beeproblem.py
import sys
sys.setrecursionlimit(1000000)

h, n, m = map(int, input().split())
grid = []
for _ in range(n):
    letters = input().split()
    new = []
    for letter in letters:
        if letter == '.':
            new .append(0)
        elif letter == '#':
            new.append(1)
        
    grid.append(new)
# print(grid)

def fill(i, j):
    if i < 0 or j < 0 or i >= n or j >= m:
        return 0
    if grid[i][j] == 1:
        return 0

    count = 1
    grid[i][j] = 1

    if i % 2 == 1:
        s = 0
        b = 2
    else:
        s = -1
        b = 1

    # top bottom
    for x in range(s, b):
        for y in [-1,1]:
            count += fill(i + y, j + x)

    # left right
    for x in [-1,1]:
        count += fill(i, j + x)

    return count

counts = []
for i in range(n):
    for j in range(m):
        tmp = fill(i, j)
        if tmp != 0:
            counts.append(tmp)

counts.sort()
# print(counts)
num = 0

while h > 0:
	h -= counts.pop()
	num += 1

print(num)
